packages/runner/src/source-code.ts

```diff
@@ -24,9 +26,69 @@ export const isESMExtension = (path: string) =>
 
 export const temporaryLoadedModules = new Map<
   string,
-  TransformFunction | null
+  { transformer: TransformFunction | null; parser?: string | null }
 >();
 
+export const getJSCodeshiftParser = async (source: string) => {
+  const getParser = async () => {
+    type Exports = NonDefaultExports | (() => void);
+
+    const hashDigest = createHash("sha256").update(source).digest("hex");
+
+    try {
+      // CJS
+      const module = { exports: {} as Exports };
+
+      const keys = ["module", "exports", "require"];
+      const values = [module, module.exports, require];
+
+      new Function(...keys, source).apply(null, values);
+
+      return typeof module.exports === "object"
+        ? module.exports.parser ?? null
+        : null;
+    } catch (err) {
+      // ESM
+      try {
+        const alreadyLoaded = temporaryLoadedModules.get(hashDigest);
+        if (alreadyLoaded?.parser) return alreadyLoaded.parser;
+
+        const tempFilePath = join(__dirname, `temp-module-${hashDigest}.mjs`);
+
+        await writeFile(tempFilePath, source);
+        const module = (await import(
+          `file://${tempFilePath}`
+        )) as NonDefaultExports;
+
+        const parser = module.parser ?? null;
+
+        temporaryLoadedModules.set(hashDigest, {
+          transformer: alreadyLoaded?.transformer ?? null,
+          parser,
+        });
+        return parser;
+      } catch (err) {
+        return null;
+      }
+    }
+  };
```

Comment:
Can we somehow avoid performing this twice? We're already doing the bulk of this in `getTransformer`?